### ISSUE TITLE
Quasar props can not contain spaces

### ISSUE THREAD

**Issue by rodja** [zauberzeug/nicegui#98]
### Discussed in https://github.com/zauberzeug/nicegui/discussions/97

<div type='discussions-op-text'>

<sup>Originally posted by **gbrandt** September 27, 2022</sup>
I am using ui.input() and it allows a prop called 'error-message'.

`self.input_add_user_email = ui.input(label='Email', on_change=self.on_change_email).props('error-message=Invalid Email').style('width: 300px')`

However only the text `Invalid` is shown on the error when I do this:  `self.button_add_user.props('disable=false')`

Is there anyway to get a string with spaces into the error message?</div>

This is indeed a limitation of our current simple property parsing:
https://github.com/zauberzeug/nicegui/blob/f417d7a64b43f00e52dd843445a8a9d8b9b366b9/nicegui/elements/element.py#L93

Task List:

- [x] write integration test (skip execution until issue is fixed)
- [x] unit-test `str_to_dict` function which is used by `element.props`
- [x] allow multi-word values in props
- [x] enable integration test if everything is running
- [ ] review

**Comment by falkoschindler** [comment]
Here is a more compact reproduction:

```python
ui.input().props('error error-message=Invalid Email')
```

A quick workaround is to set such properties directly on the JustPy view:

```python
input = ui.input().props('error')
input.view.error_message = 'Invalid Email'
```

Of course, it would be nice to allow for spaces in NiceGUI's `.props()` method. @rodja Do you already have an API in mind? Using quotes to denote strings? If so, we probably use a library for parsing the string rather than doing it ourselves.

**Comment by falkoschindler** [comment]
A simple solution would be another method for single props:

```python
.prop(error_message='Invalid Email')
```

**Comment by falkoschindler** [comment]
This might be what we need for `.props()` with spaces: https://gist.github.com/AntonFriberg/d3440c7d6f4d57d8ac64cb70b4aefd4f

**Comment by rodja** [comment]
> Do you already have an API in mind? Using quotes to denote strings? If so, we probably should use a library for parsing the string rather than doing it ourselves.

@falkoschindler yes, using quotes would be most intuitive. Adding another `.prop` method feels a bit hacky. But the gist you have found suggests the difficulties.

**Comment by falkoschindler** [comment]
Inspired by the [gist](https://gist.github.com/AntonFriberg/d3440c7d6f4d57d8ac64cb70b4aefd4f) I came up with quite simple parsers for props, styles and classes:

```python
def parse_props(text: str) -> Dict[str, str]:
    lexer = shlex.shlex(text, posix=True)
    lexer.whitespace = ' '
    lexer.wordchars += '=-'
    return dict(word.split('=', 1) if '=' in word else (word, True) for word in lexer)

def parse_style(text: str) -> Dict[str, str]:
    lexer = shlex.shlex(text, posix=True)
    lexer.whitespace = ';'
    lexer.wordchars += ':- '
    return dict(map(str.strip, word.split(':', 1)) for word in lexer)

def parse_classes(text: str) -> Set[str]:
    lexer = shlex.shlex(text, posix=True)
    lexer.whitespace = ' '
    lexer.wordchars += '-'
    return set(lexer)

print(parse_props('error error-message="Invalid & Email"'))
print(parse_style('bg-color:red; foo:bar'))
print(parse_classes('foo bar-buz one_two'))
```

Each lexer defines `whitespace` that separates consecutive key-value blocks. `wordchars` are letters, numbers and "_" by default. We the key-value separator and other allows characters like "-". Note that everything in quotes is allowed anyway, like the "&" in the props example. The last step is to process the lexing result: Props are automatically true if not value is given; and we strip whitespace from style keys and values.

These functions could replace `str_to_dict`. And the lexers could even be instantiated once and reused if performance is an issue.

**Comment by falkoschindler** [comment]
I integrated two new parse functions based on shlex. But I left `classes()` untouched, because this parser is basically a string split.

**Comment by rodja** [comment]
Released in [0.9.10](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.10)

**Comment by rodja** [comment]
Bugfix release in [0.9.11](https://github.com/zauberzeug/nicegui/releases/tag/v0.9.11)
